#
#
#	Prepare microRNA seeds file to TargetScan v7.0
#	analysis.
#
#	Author: Javier Montalvo-Arredondo.
#	Contact: 


library("Biostrings")

prepare_mirna_files_to_targetscan_pending <- function(seeds, output = "miRNA") {
	# Requires a List object with DNAStringSet
	# of all posible offset kmers of microRNAs.
	# This object is generated by transform_mature_mirna_to_seed_sequence()
	# function.

	k7mer_A1 <- reverseComplement(seeds$k7mer_A1)
	k7mer_m8 <- reverseComplement(seeds$k7mer_m8)
	k7mer_A1_table <- data.frame(name = gsub(" .*", "", names(as.character(k7mer_A1))),
					sequence = as.character(k7mer_A1))
	k7mer_m8_table <- data.frame(name = gsub(" .*", "", names(as.character(k7mer_m8))),
					sequence = as.character(k7mer_m8))
	rownames(k7mer_A1_table) <- NULL
	rownames(k7mer_m8_table) <- NULL
	k7mer_A1_table$species <- rep("1001", length(k7mer_A1[,1]))
	k7mer_m8_table$species <- rep("1001", length(k7mer_m8[,1]))
	
	write.table(k7mer_A1_table,
		    file = paste(output, "k7mer_A1.ints", sep="."),
		    quote = FALSE,
		    row.names = FALSE,
		    col.names=FALSE,
		    sep="\t")

	write.table(k7mer_m8_table,
		    file = paste(output, "k7mer_m8.ints", sep="."),
		    quote = FALSE,
		    row.names = FALSE,
		    col.names = FALSE,
		    sep="\t")
}

prepare_mirna_files_to_targetscan <- function(mirna_fasta_file, organismID = "1001", output = "out") {

	#	Reads fasta file
	seqs <- readRNAStringSet(mirna_fasta_file,
				 format = "fasta")

	#	Extracting Xenopus laevis miRNAs
	seqs <- seqs[grepl("^xla-", names(seqs))]

	#	Subsetting sequences
	subseqs <- subseq(x = seqs,
			  start = 2,
			  width = 7)

	#	Getting and stripping miRNA names
	mir_names <- names(seqs)
	mir_names <- gsub(" [[:alpha:]]*[[:digit:]]*.*", "", mir_names)

	#	Creating organism ID vector
	orgID <- rep(organismID,
		     length(mir_names))

	#	Getting sequences as a character vector.
	mature_seqs <- as.character(seqs)
	names(mature_seqs) <- NULL
	seed_seqs <- as.character(subseqs)
	names(seed_seqs) <- NULL

	#	Creating data frames
#	mirna_table_to_ts_file <- data.frame(name = mir_names,
#				       seq = seed_seqs,
#				       orgID = orgID)
	mirna_table_to_ts_file <- data.frame(name = seed_seqs,
					     seq = seed_seqs,
					     orgID = orgID)

	mirna_table_to_ts_file <- mirna_table_to_ts_file[!duplicated(mirna_table_to_ts_file),]

	mirna_table_to_tsCTS_file <- data.frame(seed_name = seed_seqs,
					  orgID = orgID,
					  name = mir_names,
					  seq = mature_seqs)

	#	Writing data frames to files
	write.table(x = mirna_table_to_ts_file,
		    file = paste(output, "ints", sep="."),
		    row.names = FALSE,
		    col.names = FALSE,
		    quote = FALSE,
		    sep = "\t")

	write.table(x = mirna_table_to_tsCTS_file,
		    file = paste(output, "intsCTS", sep="."),
		    row.names = FALSE,
		    col.names = FALSE,
		    quote = FALSE,
		    sep = "\t")

}
